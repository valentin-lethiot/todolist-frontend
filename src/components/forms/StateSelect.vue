<script setup lang="ts">
import { TaskStatus } from '@/models/tasks.entity'
import { computed } from 'vue'
import { useI18n } from 'vue-i18n'

const { t } = useI18n()

type Props = {
  modelValue: string
}
const props = defineProps<Props>()

const emit = defineEmits(['update:modelValue'])

const emitNewValue = (event: any) => {
  emit('update:modelValue', event.target.value)
}

const backgroundColor = computed(() => {
  switch (props.modelValue) {
    case TaskStatus.TODO:
      return 'bg-red-600'
    case TaskStatus.DOING:
      return 'bg-orange-600'
    case TaskStatus.DONE:
      return 'bg-green-600'
    default:
      return 'bg-red-600'
  }
})
</script>

<template>
  <select
    :value="modelValue"
    :class="[
      'border text-sm rounded-lg p-1 border-gray-60 text-white hover:cursor-pointer',
      backgroundColor,
    ]"
    @change="emitNewValue"
  >
    <option :value="TaskStatus.TODO" selected class="bg-red-500">{{ t('status.todo') }}</option>
    <option :value="TaskStatus.DOING" class="bg-orange-500">{{ t('status.doing') }}</option>
    <option :value="TaskStatus.DONE" class="bg-green-500">{{ t('status.done') }}</option>
  </select>
</template>
