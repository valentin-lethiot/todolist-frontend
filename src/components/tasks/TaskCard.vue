<script setup lang="ts">
import { Task, TaskStatus } from '@/models/tasks.entity'
import TaskCheckbox from '../forms/TaskCheckbox.vue'
import StateSelect from '../forms/StateSelect.vue'
import EditIcon from '../ui/EditIcon.vue'
import { useTasksStore } from '@/stores/tasks.store'

const { updateTask } = useTasksStore()

type Props = {
  task: Task
}
const { task } = defineProps<Props>()

const emit = defineEmits(['editTask'])

const emitEditTask = () => emit('editTask')

const updateTaskStatus = (status: TaskStatus) => {
  updateTask(task.id!, { ...task, status: status as TaskStatus })
}
</script>

<template>
  <div class="rounded-xl w-100 p-4 bg-gray-800 border-gray-700">
    <div class="flex flex-col">
      <div class="flex justify-between">
        <p class="w-5/6 text-lg font-bold text-white break-all">
          {{ task.title }}
        </p>
        <div>
          <button
            class="rounded-full p-2 hover:bg-gray-600 text-lg font-bold text-white hover:cursor-pointer"
            @click="emitEditTask"
          >
            <EditIcon />
          </button>
        </div>
      </div>
      <p class="text-gray-300">
        {{ task.description }}
      </p>
      <div class="flex mt-2 justify-end">
        <StateSelect label="Status" v-model="task.status" @update:model-value="updateTaskStatus" />
      </div>
    </div>
  </div>
</template>
