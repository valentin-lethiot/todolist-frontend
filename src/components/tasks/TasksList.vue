<script setup lang="ts">
import { Task } from '@/models/tasks.entity'
import TaskCard from '@/components/tasks/TaskCard.vue'

type Props = {
  tasks: Task[]
}
defineProps<Props>()
</script>

<template>
  <div v-for="task in tasks" :key="task.id" class="flex flex-col my-1">
    <TaskCard :task="task" />
  </div>
</template>
