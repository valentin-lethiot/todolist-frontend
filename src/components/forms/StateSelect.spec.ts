import { describe, it, expect, beforeEach } from 'vitest'
import { mount, VueWrapper } from '@vue/test-utils'
import StateSelect from './StateSelect.vue'
import { TaskStatus } from '@/models/tasks.entity'

describe('StateSelect', () => {
  let wrapper: VueWrapper

  const defaultProps = {
    modelValue: TaskStatus.TODO,
  }

  beforeEach(() => {
    wrapper = mount(StateSelect, {
      props: defaultProps,
    })
  })

  describe('renders', () => {
    it('should renders with TODO status', () => {
      wrapper = mount(StateSelect, {
        props: {
          modelValue: TaskStatus.TODO,
        },
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders properly with DOING status', () => {
      wrapper = mount(StateSelect, {
        props: {
          modelValue: TaskStatus.DOING,
        },
      })

      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders properly with DONE status', () => {
      wrapper = mount(StateSelect, {
        props: {
          modelValue: TaskStatus.DONE,
        },
      })

      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  it('emits update:modelValue event when selection changes', async () => {
    await wrapper.find('select').setValue(TaskStatus.DOING)

    expect(wrapper.emitted('update:modelValue')).toBeTruthy()
    expect(wrapper.emitted('update:modelValue')![0]).toEqual([TaskStatus.DOING])
  })
})
