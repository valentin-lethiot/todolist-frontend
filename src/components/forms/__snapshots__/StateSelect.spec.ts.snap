// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StateSelect > renders > renders properly with DOING status 1`] = `
"<select class="border text-sm rounded-lg p-1 border-gray-60 text-white hover:cursor-pointer bg-orange-600" value="DOING">
  <option selected="" class="bg-red-500" value="TODO">A faire</option>
  <option class="bg-orange-500" value="DOING">En cours</option>
  <option class="bg-green-500" value="DONE">Terminé</option>
</select>"
`;

exports[`StateSelect > renders > renders properly with DONE status 1`] = `
"<select class="border text-sm rounded-lg p-1 border-gray-60 text-white hover:cursor-pointer bg-green-600" value="DONE">
  <option selected="" class="bg-red-500" value="TODO">A faire</option>
  <option class="bg-orange-500" value="DOING">En cours</option>
  <option class="bg-green-500" value="DONE">Terminé</option>
</select>"
`;

exports[`StateSelect > renders > should renders with TODO status 1`] = `
"<select class="border text-sm rounded-lg p-1 border-gray-60 text-white hover:cursor-pointer bg-red-600" value="TODO">
  <option selected="" class="bg-red-500" value="TODO">A faire</option>
  <option class="bg-orange-500" value="DOING">En cours</option>
  <option class="bg-green-500" value="DONE">Terminé</option>
</select>"
`;
